storage/entry_cache.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2014 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Kevin GuanJian (keynovo@gmail.com)
+
+package storage
+
+import (
+	"sync"
+
+	"github.com/biogo/store/llrb"
+	"github.com/cockroachdb/cockroach/util/cache"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+type entryCacheKey uint64
+
+// Compare implements the llrb.Comparable interface for entryCacheKey, so that
+// it can be used as a key for util.OrderedCache.
+func (a entryCacheKey) Compare(b llrb.Comparable) int {
+	if a > b.(entryCacheKey) {
+		return 1
+	} else if a < b.(entryCacheKey) {
+		return -1
+	}
+	return 0
+}
+
+// A raftEntryCache maintains the recent entries
+type raftEntryCache struct {
+	// entryCache caches recent entries, the Index as the key.
+	entryCache *cache.OrderedCache
+	// the minimum Index in the cache, evicted when cache is filled.
+	firstIndex uint64
```

Comment:
Can't you use `OrderedCache.Ceil(entryCacheKey(0))` instead of tracking `firstIndex` yourself? Or, since you're tracking `firstIndex` yourself, it should probably be more efficient to use an `UnorderedCache`.